# Assignment 5 – Prescriptive Analytics: Linear Optimisation

In [35]:
from datetime import datetime, timedelta
print(f'Updated {datetime.now().strftime("%d.%m.%Y")} / Tatu Erkinjuntti')

Updated 25.11.2025 / Tatu Erkinjuntti


## Learning Objectives

This assignment introduces linear optimisation and how to use optimisation in decision-making. In doing so, you will also learn how to present information clearly and how to document and explain calculation steps created in notebooks.



## Initial information

The company manufactures two products, a grate and a stirrer. The grate has a margin of €250 and the stirrer a margin of €320 per product. There are four stages in the manufacturing process: cutting, punching, pressing, and painting.

Each work stage requires the following amount of time (minutes) from the products:

<table align="center">
<tr>
<th></th>
<th>cutting</th>
<th>punching</th>
<th>pressing</th>
<th>painting</th>
</tr>

<tr>
<th>grate</th>
<td>12</td>
<td>16</td>
<td>12</td>
<td>22</td>
</tr>

<tr>
<th>stirrer</th>
<td>20</td>
<td>7</td>
<td>6</td>
<td>15</td>
</tr>

<tr>
<th>capasity</th>
<td>25&nbsp;000</td>
<td>30&nbsp;000</td>
<td>30&nbsp;000</td>
<td>20&nbsp;000</td>
</tr>

</table>

## Matters to be found out

Find out:

1. Optimal production program. (20 %)
2. Maximum margin. (20 %)
3. The company has €25,000 at its disposal, which it can use to increase capacity at any stage of production, but only at one stage, at a cost of €1/min. How should the company use the €25,000 to increase capacity? (30 %)
4. In relation to the initial situation (the additional capacity of €25,000 is not used in this case), the company still has a product called jiggery-pokery, which has a margin of €350 per product and requires the following capacity:
<table>
<tr>
<th></th>
<th>cutting</th>
<th>punching</th>
<th>pressing</th>
<th>painting</th>
</tr>
<tr>
<th>jiggery-pokery</th>
<td>30</td>
<td>20</td>
<td>10</td>
<td>25</td>
</tr>
</table>
What is the company's optimal production and corrsponding maximum profit margin?  (30 %)

In [36]:
# Before we can begin, import necessary libraries.

from pulp import LpMaximize, LpProblem, LpStatus, lpSum, LpVariable

### Assignment part 1 and 2

We start off by finding out the optimal production program for the grates and stirrers and based on these, determine the maximum margin.

In [37]:
# Create the model
model = LpProblem(name="optimal-production", sense=LpMaximize)

# Define variables (the number of grates and stirrers to produce)
grate_procution = LpVariable(name="Grate", lowBound=0)
stirrer_production = LpVariable(name="Stirrer", lowBound=0)

# Define the constraints based on each production stage
model += (12 * grate_procution + 20 * stirrer_production <= 25000, "cutting_capacity")
model += (16 * grate_procution + 7 * stirrer_production <= 30000, "punching_capacity")
model += (12 * grate_procution + 6 * stirrer_production <= 30000, "pressing_capacity")
model += (22 * grate_procution + 15 * stirrer_production <= 20000, "painting_capacity")

# Define the objective function (Maximize margin)
model += lpSum([250 * grate_procution, 320 * stirrer_production])


Lets check the model.

In [38]:
model

optimal-production:
MAXIMIZE
250*Grate + 320*Stirrer + 0.0
SUBJECT TO
cutting_Capacity: 12 Grate + 20 Stirrer <= 25000

punching_Capacity: 16 Grate + 7 Stirrer <= 30000

pressing_Capacity: 12 Grate + 6 Stirrer <= 30000

painting_Capacity: 22 Grate + 15 Stirrer <= 20000

VARIABLES
Grate Continuous
Stirrer Continuous

**NOTE:** since the assignment does not specify to use integer optimisation (Integer Linear Programming), we will stick to the default continuous optimisation.

Lets optimise.

In [39]:
model.solve()

Welcome to the CBC MILP Solver 
Version: 2.10.3 
Build Date: Dec 15 2019 

command line - /Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/pulp/apis/../solverdir/cbc/osx/i64/cbc /var/folders/3j/zrw9gdxj61g3x4ypt77qk04h0000gp/T/43abeb684f1b45cb9f62f09c629fc89c-pulp.mps -max -timeMode elapsed -branch -printingOptions all -solution /var/folders/3j/zrw9gdxj61g3x4ypt77qk04h0000gp/T/43abeb684f1b45cb9f62f09c629fc89c-pulp.sol (default strategy 1)
At line 2 NAME          MODEL
At line 3 ROWS
At line 9 COLUMNS
At line 20 RHS
At line 25 BOUNDS
At line 26 ENDATA
Problem MODEL has 4 rows, 2 columns and 8 elements
Coin0008I MODEL read with 0 errors
Option for timeMode changed from cpu to elapsed
Presolve 3 (-1) rows, 2 (0) columns and 6 (-2) elements
0  Obj -0 Dual inf 570 (2)
2  Obj 405576.92
Optimal - objective value 405576.92
After Postsolve, objective 405576.92, infeasibilities - dual 0 (0), primal 0 (0)
Optimal objective 405576.9231 - 2 iterations time 0.002, Pres

1

Lets check the results.

In [40]:
# Status

print(f"status: {model.status}, {LpStatus[model.status]}")

status: 1, Optimal


In [41]:
for var in model.variables():
    print(f"Optimal number for {var.name} production is {var.value():.0f}")

Optimal number for Grate production is 96
Optimal number for Stirrer production is 1192


In [42]:
print(f"Maximum Margin is {model.objective.value():,.2f} €")

Maximum Margin is 405,576.93 €


#### Assignment part 1 and 2 summary

By using continuous optimisation we can determine that the optimal number for grate production is 96 and for the stirrers this is 1192.

Maximum margin is 405,576.93 €

### Assignment part 3